src/providers/cryptoauthlib/aead.rs

```diff
@@ -0,0 +1,155 @@
+    )
+}
+
+impl Provider {
+    pub(super) fn psa_aead_encrypt_internal(
+        &self,
+        app_name: ApplicationName,
+        op: psa_aead_encrypt::Operation,
+    ) -> Result<psa_aead_encrypt::Result> {
+        match get_tag_length(&op.alg) {
+            Some(tag_length) => {
+                let key_triple =
+                    KeyTriple::new(app_name, ProviderId::CryptoAuthLib, op.key_name.clone());
+                let key_id = self.key_info_store.get_key_id::<u8>(&key_triple)?;
+                let key_attributes = self.key_info_store.get_key_attributes(&key_triple)?;
+                op.validate(key_attributes)?;
+
+                let aead_param_gcm = rust_cryptoauthlib::AeadParam {
+                    nonce: op.nonce.to_vec(),
+                    tag_length: Some(tag_length as u8),
+                    additional_data: Some(op.additional_data.to_vec()),
+                    ..Default::default()
+                };
+
+                let aead_param_ccc = aead_param_gcm.clone();
+                let mut aead_algorithm = rust_cryptoauthlib::AeadAlgorithm::Gcm(aead_param_gcm);
+                if is_ccm_selected(&op.alg) {
+                    aead_algorithm = rust_cryptoauthlib::AeadAlgorithm::Ccm(aead_param_ccc);
+                }
+
+                let mut plaintext = op.plaintext.to_vec();
+
+                match self
+                    .device
+                    .aead_encrypt(aead_algorithm, key_id, &mut plaintext)
+                {
+                    Ok(tag) => {
+                        plaintext.extend(tag);
+
+                        Ok(psa_aead_encrypt::Result {
+                            ciphertext: plaintext.into(),
+                        })
+                    }
+                    Err(error) => {
+                        error!("aead_encrypt failed CAL error {}.", error);
+                        Err(ResponseStatus::PsaErrorGenericError)
+                    }
+                }
+            }
+            None => {
+                error!("aead_encrypt failed, algorithm not supported");
+                Err(ResponseStatus::PsaErrorNotSupported)
+            }
+        }
+    }
+
+    pub(super) fn psa_aead_decrypt_internal(
+        &self,
+        app_name: ApplicationName,
+        op: psa_aead_decrypt::Operation,
+    ) -> Result<psa_aead_decrypt::Result> {
+        match get_tag_length(&op.alg) {
+            Some(tag_length) => {
+                let key_triple =
+                    KeyTriple::new(app_name, ProviderId::CryptoAuthLib, op.key_name.clone());
+                let key_id = self.key_info_store.get_key_id::<u8>(&key_triple)?;
+                let key_attributes = self.key_info_store.get_key_attributes(&key_triple)?;
+                op.validate(key_attributes)?;
+
+                let mut ciphertext: Vec<_> = op.ciphertext.to_vec();
+                let tag: Vec<_> = ciphertext
+                    .drain((ciphertext.len() - tag_length)..)
+                    .collect();
```

Comment:
Maybe [`split_off`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off) can be used in an easier way here?